### PR TITLE
Adds new methods for safely adding path parts to an URL builder

### PR DESCRIPTION
**Pull request by sabieber**
Fixes: [SIRI-641](https://scireum.myjetbrains.com/youtrack/issue/SIRI-641)

### CODE REVIEW

File: src/main/java/sirius/kernel/commons/URLBuilder.java [hunk 1]
```diff
@@ -70,6 +74,55 @@ public URLBuilder(@Nonnull String protocol, @Nonnull String host) {
         url.append(host);
     }
 
+    /**
+     * Adds a path part to the url.
```
Comment (jakobvogel) [inline src/main/java/sirius/kernel/commons/URLBuilder.java:78]: Sollten wir das als Abkürzung nicht groß schreiben?

```suggestion
     * Adds a path part to the URL.
```

File: src/main/java/sirius/kernel/commons/URLBuilder.java [hunk 2]
```diff
@@ -70,6 +74,55 @@ public URLBuilder(@Nonnull String protocol, @Nonnull String host) {
...
+     * <p>
+     * Once the first parameter has been added, the path can no longer be modified. Also, the part itself can (but
+     * shouldn't) contain parameters, usually led by an initial question mark.
+     *
+     * @param uriPartToAdd the uri part to add. Included trailing or leading '/' are handled automatically.
```
Comment (jakobvogel) [inline src/main/java/sirius/kernel/commons/URLBuilder.java:83]: Dto.

```suggestion
     * @param uriPartToAdd the URI part to add. Included trailing or leading '/' are handled automatically.
```

File: src/main/java/sirius/kernel/commons/URLBuilder.java [hunk 3]
```diff
@@ -70,6 +74,55 @@ public URLBuilder(@Nonnull String protocol, @Nonnull String host) {
...
+     *                     If this contains a '?' to add parameters, no more parts can be added afterwards.
+     * @return the builder itself for fluent method calls
+     */
+    public URLBuilder addSafePart(@Nullable String uriPartToAdd) {
+        if (Strings.isEmpty(uriPartToAdd)) {
+            return this;
+        }
+
+        if (questionMark.isToggled()) {
+            throw new IllegalStateException(Strings.apply("Cannot add '%s'! Parameters where already added to: '%s'.",
```
Comment (idlira) [inline src/main/java/sirius/kernel/commons/URLBuilder.java:93]:
```suggestion
            throw new IllegalStateException(Strings.apply("Cannot add '%s'! Parameters were already added to: '%s'.",
```

**Review by idlira — APPROVED** [review]

**Review by jakobvogel — APPROVED** [review]

**Comment by sabieber** [comment]
Thanks for your suggestions @idlira & @jakobvogel but had to apply them to the whole class as it was wrong everywhere and I just oriented myself at the other existing methods 😉

**Review by idlira — APPROVED** [review]

**Review by andyHa — APPROVED** [review]

**Comment by qw3ry** [comment]
If we want to remove `addPart`, why not just make the PR breaking and change its semantics? Once `addPart` is gone, the remaining `addSafePart` will sound somewhat awkward, and clumsy (without an unsafe counterpart)

**Comment by sabieber** [comment]
> If we want to remove `addPart`, why not just make the PR breaking and change its semantics? Once `addPart` is gone, the remaining `addSafePart` will sound somewhat awkward, and clumsy (without an unsafe counterpart)

The idea was that it is much easier and safer to detect where the old method is called with semantics that are no longer supported (as seen in the tests or here: https://github.com/scireum/oxomi/pull/9061/files#diff-7be4d3f73bf8a93ed7740720be6f70c780d76dbdaa5e11713dbc619b096ab1a8R255)

Also it was discussed to maybe at a later date add a `addRawPart` if needed, that skips all the safety checks.

**Review by qw3ry — APPROVED** [review]

**Review by mkeckmkeck — APPROVED** [review]

**Pull request merged**